hspec-core/src/Test/Hspec/Core/Runner/Eval.hs

```diff
@@ -53,22 +53,27 @@ every seconds action = do
 
 type FormatResult = Either E.SomeException Result -> FormatM ()
 
-parallelize :: Bool -> (ProgressCallback -> IO Result) -> ProgressCallback -> FormatResult -> IO (FormatM ())
-parallelize isParallelizable e
-  | isParallelizable = runParallel e
-  | otherwise = runSequentially e
+parallelize :: Maybe QSem -> Bool -> (ProgressCallback -> IO Result) -> ProgressCallback -> FormatResult -> IO (FormatM ())
+parallelize jobsSem isParallelizable e
+  | isParallelizable = runParallel jobsSem e
+  | otherwise = runSequentially jobsSem e
 
-runSequentially :: (ProgressCallback -> IO Result) -> ProgressCallback -> FormatResult -> IO (FormatM ())
-runSequentially e reportProgress formatResult = return $ do
-  result <- liftIO $ evalExample (e reportProgress)
+limitParallelism :: Maybe QSem -> IO a -> IO a
+limitParallelism jobsSem = case jobsSem of
+  Nothing -> id
+  Just s -> E.bracket_ (waitQSem s) (signalQSem s)
+
+runSequentially :: Maybe QSem -> (ProgressCallback -> IO Result) -> ProgressCallback -> FormatResult -> IO (FormatM ())
+runSequentially jobsSem e reportProgress formatResult = return $ do
+  result <- liftIO $ limitParallelism jobsSem $ evalExample (e reportProgress)
```

Comment:
My reasoning is that the main thread, when producing the spec output, will eventually try to take the result of a parallelizable example from an `MVar` and block on that operation.  So the main thread would need to signal before it prints the result of the first parallelizable example, I guess.  But after the first parallelizable example there may still be more sequential examples, so proceeding naively we would end up with one more thread than specified again.

I imagine it may be possible to get it "right" by using wait/signal multiple times in the main thread, but I'm not sure the added complexity is worth it.